### PR TITLE
use arrrow keys for address selection of distance expense

### PR DESCRIPTION
**Pull request by sourcecodedeveloper**
### Fixed Issues

$ https://github.com/Expensify/App/issues/34350
PROPOSAL: https://github.com/Expensify/App/issues/34350#issuecomment-1898112442

### Tests

1. Go to workspace chat and create distance expense
2. Enter start address
3. Verify the arrow keys can be used to navigate through the results.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
N/A

</details>

<details>
<summary>Android: mWeb Chrome</summary>
N/A

</details>

<details>
<summary>iOS: Native</summary>
N/A

</details>

<details>
<summary>iOS: mWeb Safari</summary>
N/A

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

[web.webm](https://github.com/Expensify/App/assets/141340971/93ae99f6-7b2c-4bf2-bf62-3c90107a7a1c)

</details>

<details>
<summary>MacOS: Desktop</summary>

[desktop.webm](https://github.com/Expensify/App/assets/141340971/2bf8275c-ba65-4bec-bde1-362a1e6d2299)

</details>

### CODE REVIEW

File: src/components/AddressSearch/index.tsx [hunk 1]
```diff
@@ -224,6 +233,30 @@ function AddressSearch(
 
         onPress?.(values);
     };
+    const selectFocusedOption = () => {
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:236]: Let's memoize this with `useCallback` so `useKeyboardShortcut` doesn't add and remove an event listener on every render.

File: src/components/AddressSearch/index.tsx [hunk 2]
```diff
@@ -224,6 +233,30 @@ function AddressSearch(
...
+
+        // After we select an option, we set displayListViewBorder to false to prevent UI flickering
+        setDisplayListViewBorder(false);
+        setIsFocused(false);
+
+        // Clear location error code after address is selected
+        setLocationErrorCode(null);
+    };
+    useKeyboardShortcut(CONST.KEYBOARD_SHORTCUTS.ENTER, selectFocusedOption, {
+        captureOnInputs: true,
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:255]: Most of these config values are the default. The only one that's different is `shouldBubble`. Can you first try removing this config object entirely and see if that works? If not, let's pare this back to include only `shouldBubble: true`.

File: src/components/AddressSearch/index.tsx [hunk 3]
```diff
@@ -343,14 +376,21 @@ function AddressSearch(
                         listEmptyComponent={listEmptyComponent}
                         listLoaderComponent={listLoader}
                         renderHeaderComponent={renderHeaderComponent}
-                        renderRow={(data) => {
+                        renderRow={(data, index) => {
                             const title = data.isPredefinedPlace ? data.name : data.structured_formatting.main_text;
                             const subtitle = data.isPredefinedPlace ? data.description : data.structured_formatting.secondary_text;
+                            if (focusedIndex === index) {
+                                setFocusedIndex(index);
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:383]: I don't think this should be necessary. `setFocusedIndex` is mostly returned from `useArrowKeyFocusManager` if you need to reset it to `-1`

File: src/components/AddressSearch/index.tsx [hunk 4]
```diff
@@ -343,14 +376,21 @@ function AddressSearch(
...
+                            }
                             return (
-                                <View>
-                                    {title && <Text style={[styles.googleSearchText]}>{title}</Text>}
-                                    <Text style={[styles.textLabelSupporting]}>{subtitle}</Text>
-                                </View>
+                                <Hoverable>
+                                    {(hovered) => (
+                                        <View style={[styles.pv4, styles.ph3, StyleUtils.getBackgroundAndBorderStyle(hovered || focusedIndex === index ? theme.border : theme.appBG)]}>
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:388]: **question:** Why did move the padding from where it was to this view?

File: src/components/AddressSearch/index.tsx [hunk 5]
```diff
@@ -343,14 +376,21 @@ function AddressSearch(
...
+                            if (focusedIndex === index) {
+                                setFocusedIndex(index);
+                            }
                             return (
-                                <View>
-                                    {title && <Text style={[styles.googleSearchText]}>{title}</Text>}
-                                    <Text style={[styles.textLabelSupporting]}>{subtitle}</Text>
-                                </View>
+                                <Hoverable>
+                                    {(hovered) => (
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:387]:
```suggestion
                                    {(isHovered) => (
```

File: src/components/AddressSearch/index.tsx [hunk 6]
```diff
@@ -412,18 +452,22 @@ function AddressSearch(
...
                         numberOfLines={2}
                         isRowScrollable={false}
                         listHoverColor={theme.border}
                         listUnderlayColor={theme.buttonPressedBG}
-                        onLayout={(event: LayoutChangeEvent) => {
+                        onLayout={(event) => {
                             // We use the height of the element to determine if we should hide the border of the listView dropdown
                             // to prevent a lingering border when there are no address suggestions.
                             setDisplayListViewBorder(event.nativeEvent.layout.height > variables.googleEmptyListViewHeight);
+                            const {target} = event.nativeEvent as unknown as BaseSyntheticEvent;
```
Comment (roryabraham) [inline src/components/AddressSearch/index.tsx:467]: I think there's probably a better way to fix this type. I see `event.target` is valid, it's just missing the `getScrollResponser()` function. Maybe need to use generics like `LayoutChangeEvent<GooglePlacesAutoComplete>` or something

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/components/AddressSearch/index.tsx (hunk 4 above)
Comment (sourcecodedeveloper) [reply src/components/AddressSearch/index.tsx:388]: the selected row was only highlighting the title of the address and the padding region outside the title was not getting highlighted. After removing the padding, the padded region outside the title was also getting highlighted on arrow key selection.

File: src/components/AddressSearch/index.tsx (hunk 6 above)
Comment (sourcecodedeveloper) [reply src/components/AddressSearch/index.tsx:462]: I tried LayoutChangeEvent<GooglePlacesAutoComplete> but getting error Type 'LayoutChangeEvent' is not generic. The event type LayoutChangeEvent is defined as NativeSyntheticEvent<{layout: LayoutRectangle}>. So most probably the event.nativeEvent would need to be casted to BaseSyntheticEvent to get to the target.

**Comment by sourcecodedeveloper** [comment]
@roryabraham PR ready for another review.

File: src/components/AddressSearch/index.tsx (hunk 6 above)
Comment (sourcecodedeveloper) [reply src/components/AddressSearch/index.tsx:462]: I tried LayoutChangeEvent but getting error Type 'LayoutChangeEvent' is not generic. The event type LayoutChangeEvent is defined as NativeSyntheticEvent<{layout: LayoutRectangle}>. So the event.nativeEvent would need to be casted to BaseSyntheticEvent to get to the target prop.

**Comment by roryabraham** [comment]
Awaiting C+ review from @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
https://github.com/Expensify/App/assets/46707890/6467e713-b15b-4bab-86e6-816e4e685475

@sourcecodedeveloper I am seeing an issue where if we just open the start point which lists initial dropdown, at this point the arrow keys are still not working. 
Only after I type in the search or interact with dropdown, the arrow keys are working.

And I also think this is a valid comment, we should check some way of solving this.

**Comment by sourcecodedeveloper** [comment]
@abdulrahuman5196 I have the resolved the focus issue. PR is ready for another review.

**Comment by sourcecodedeveloper** [comment]
@abdulrahuman5196 I have resolved the focus issue. PR is ready for another review.

File: src/components/AddressSearch/index.tsx (hunk 6 above)
Comment (abdulrahuman5196) [reply src/components/AddressSearch/index.tsx:469]: @sourcecodedeveloper Sorry, I added this comment in last Review, but it didn't come up. 

Anyways, I think the concern here is valid, the type of `const {target}` (target) is resolving `any` in this place which would be bad representation in typescript.

Comment (sourcecodedeveloper) [reply src/components/AddressSearch/index.tsx]: @abdulrahuman5196 I was able to leverage generics to get type for target prop and not use any.

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
@sourcecodedeveloper  the arrow movement is not consistently working now. Sometimes its working and sometimes its not. 


https://github.com/Expensify/App/assets/46707890/015c74ea-392f-4b40-9523-8f2ecc69e3a5

**Comment by sourcecodedeveloper** [comment]
@abdulrahuman5196 I suggest to use on screen keyboard because I cannot tell from the video if the arrow key on the keyboard was pressed or not. The onscreen keyboard would clearly indicate when the arrow key is pressed and if the row is focused or not.

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 I suggest to use on screen keyboard because I cannot tell from the video if the arrow key on the keyboard was pressed or not. The onscreen keyboard would clearly indicate when the arrow key is pressed and if the row is focused or not.

Will provide next time onwards. Is this working for you now?

**Comment by sourcecodedeveloper** [comment]
@abdulrahuman5196 I have done some more testing and the arrow keys are functioning as expected. The only issue I found was the arrow key would get stuck when the maxIndex would get set to -1 when no results were present.

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
https://github.com/Expensify/App/assets/46707890/f9eb55d6-5527-442e-9065-bb41526cd906

@sourcecodedeveloper Still I am seeing occasional issues. Attached video along with onscreen keyboard.

**Comment by abdulrahuman5196** [comment]
@sourcecodedeveloper  I took the latest commit while facing this issue. And another issue is that the arrows are not moving to the `Use current location ` item.

**Comment by sourcecodedeveloper** [comment]
@abdulrahuman5196 PR is ready for review.

**Pull request closed without merging**